sound/soc/sof/ipc4-topology.c

```diff
@@ -1667,14 +1676,31 @@ static int sof_ipc4_prepare_process_module(struct snd_sof_widget *swidget,
 	cfg += sizeof(struct sof_ipc4_base_module_cfg);
 
 	/* copy output format to configure data payload */
-	if (process->payload_with_output_fmt) {
+	if (!fw_module->ext_library && process->payload_with_output_fmt) {
 		memcpy(cfg, &process->output_format, sizeof(struct sof_ipc4_audio_format));
 		cfg += sizeof(struct sof_ipc4_audio_format);
 	}
 
 	control_data = sof_ipc4_get_widget_init_byte_kcontrol(swidget);
-	if (control_data)
+	if (control_data) {
+		if (fw_module->ext_library) {
+			struct sof_ipc4_base_module_cfg_ext *base_ext;
+			int size;
+
+			base_ext = (struct sof_ipc4_base_module_cfg_ext *)control_data->data->data;
+			size = sizeof(struct sof_ipc4_base_module_cfg_ext) +
+								base_ext->priv_param_len +
+								(base_ext->num_sink_pin_fmts +
+								base_ext->num_source_pin_fmts) *
+								sizeof(struct sof_ipc4_pin_format);
+			if (size != control_data->data->size) {
+				dev_err(sdev->dev, "Insufficient data for widget %s initialization\n",
+					swidget->widget->name);
+				return -EINVAL;
+			}
+		}
 		memcpy(cfg, control_data->data->data, control_data->data->size);
+	}
```

Comment:
and that should be a different patch where you add the case of the external loadable libraries. Mixing the two is confusing.